napari/_vispy/layers/vectors.py

```diff
@@ -107,25 +112,231 @@ def generate_vector_meshes_2D(vectors, width, length, p=(0, 0, 1)):
+        [4,5,6],                # vector 1,   triangle i=2
+        [5,6,7],                # vector 1,   triangle i=3
+
+        ...,
+
+        [2i, 2i + 1, 2i + 2],   # vector k-1, triangle i=2k-2 (i%2=0)
+        [2i - 1, 2i, 2i + 1]    # vector k-1, triangle i=2k-1 (i%2=1)
+    ]
+    """
+    nvectors, _, ndim = vectors.shape
+
+    vectors_starts = vectors[:, 0]
+    vectors_ends = vectors_starts + length * vectors[:, 1]
+
+    vertices = np.zeros((4 * nvectors, ndim))
+    offsets = segment_normal(vectors_starts, vectors_ends, p=p)
     offsets = np.repeat(offsets, 4, axis=0)
+
     signs = np.ones((len(offsets), ndim))
     signs[::2] = -1
     offsets = offsets * signs
 
-    vertices = centers + width * offsets / 2
+    vertices[::4] = vectors_starts
+    vertices[1::4] = vectors_starts
+    vertices[2::4] = vectors_ends
+    vertices[3::4] = vectors_ends
+
+    vertices = vertices + width * offsets / 2
+
     triangles = np.array(
         [
             [2 * i, 2 * i + 1, 2 * i + 2]
             if i % 2 == 0
             else [2 * i - 1, 2 * i, 2 * i + 1]
-            for i in range(len(vectors))
+            for i in range(2 * nvectors)
+        ]
+    ).astype(np.uint32)
+
+    return vertices, triangles
+
+
+def generate_meshes_triangle_2D(vectors, width, length, p):
+    """Generates list of mesh vertices and triangles from a list of vectors.
+
+    Vectors are composed of 3 vertices and 1 triangles.
+    Vertices are generated according to the following scheme:
+    1---x---0
+     .     .
+      .   .
+       . .
+        2
+
+
+    Where x marks the start point of the vector, and the vertex 2 its end
+    point.
+
+    In the case of k 2D vectors, the output 'triangles' is:
+    [
+        [0,1,2],                # vector 0,   triangle i=0
+        [3,4,5],                # vector 1,   triangle i=1
+
+        ...,
+
+        [3i, 3i + 1, 3i + 2]    # vector k-1, triangle i=k-1
+    ]
+    """
+    nvectors, _, ndim = vectors.shape
+
+    vectors_starts = vectors[:, 0]
+    vectors_ends = vectors_starts + length * vectors[:, 1]
+
+    vertices = np.zeros((3 * nvectors, ndim))
+    offsets = segment_normal(vectors_starts, vectors_ends, p=p)
+    offsets = np.repeat(offsets, 3, axis=0)
+
+    signs = np.ones((len(offsets), ndim))
+    signs[::3] = -1
+    multipliers = np.ones((len(offsets), ndim))
+    multipliers[2::3] = 0
+
+    # here 'multipliers' is used to prevent vertex 2 from being offset
+    offsets = offsets * signs * multipliers
+
+    vertices[::3] = vectors_starts
+    vertices[1::3] = vectors_starts
+    vertices[2::3] = vectors_ends
+
+    vertices = vertices + width * offsets / 2
+
+    # faster than using the formula in the docstring
+    triangles = np.arange(3 * nvectors).reshape((-1, 3)).astype(np.uint32)
+
+    return vertices, triangles
+
+
+def generate_meshes_arrow_2D(vectors, width, length, p):
+    """Generates list of mesh vertices and triangles from a list of vectors.
```

Comment:
Could you update the docstrings to conform with https://numpydoc.readthedocs.io/en/latest/format.html? Your descriptions are excellent but it would be great to have the argument list — I don't know what p does without delving into the code! 😊 

It's also unclear what length means for the vectors — isn't that implied in the data? I see that it's multiplied by the second component of the vector data but I don't understand why...

Reply:
No problem. When I started working on this, only `generate_vector_meshes` was properly formatted; `generate_vector_meshes_2D` had only a single-line description, so I thought this was a choice.

Reply:
Absolutely!
I just realized that many docstrings still had "Returns" fields with 
`vertices : (4N, 2) array for 2D and (8N, 2) array for 3D`
which I forgot to update... 

Would something like 
`vertices : (kN, 2) array for 2D and (2kN, 2) array for 3D, with k=4, 3, or 7 for vector_style='line', 'triangle', or 'arrow' respectively`
be appropriate ?

Reply:
Regarding your comment about the `length` parameter: maybe it has been ill-named. The actual physical lengths of vectors are indeed directly given by the magnitude of the second component of the vector data, but the `length` parameter is an extra multiplicative factor that is applied to all vectors at the same time to allow the user to scale them in the viewer (e.g when the physical lengths are too small and are not directly visible on the viewer).

Maybe `scaling_factor` or `vector_scale` would be a better name ?

Reply:
Last comment about the `length` parameter: right now, if the "length changed" event is triggered (same for the `width` parameter by the way), `test_generate_vector_meshes` is launched again so that every meshes and triangles are computed again, which is highly ineffective as only the positions of a few vertices would actually need to be multiplied by a scalar.

Maybe a next PR could tackle this optimization issue.

Reply:
Yeah, I always doung `length` was a bad name, but it's pretty central and user-facing, so if we change it we should introduce deprecations (shouln't be too hard to do, happy to help here if you need it). I like `scaling_factor`, or ever more explicitly `length_multiplier` or `length_scaling`. @jni?

Reply:
Thanks @brisvag, this could be another interesting project for a PR. If you have any references in the current napari codebase in which a parameter's name is being replaced (and deprecation notices are implemented), I could start taking a look at it.